### PR TITLE
Watches for resources with multiple ownership

### PR DESCRIPTION
**Pull request by Teneroy**
**Description**

Changes proposed in this pull request:

- Increased number of watched resources for logpipeline
- Allowed multiple ownership for a resource
- Adapted tests for multiple ownership
- Fixed bugs related to infinite reconciliation for tracepipelines and metricpipelines

**Related issue(s)**

#107

### CODE REVIEW

**Comment by Teneroy** [comment]
retest

File: controllers/telemetry/logpipeline_controller.go [hunk 1]
```diff
@@ -60,16 +61,18 @@ func (r *LogPipelineReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 func (r *LogPipelineReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&telemetryv1alpha1.LogPipeline{}).
+		Owns(&corev1.Secret{}).
```
Comment (chrkl) [inline controllers/telemetry/logpipeline_controller.go:64]: Replace the `Owns` calls with `Watches` as it is done [here](https://github.com/kyma-project/telemetry-manager/blob/main/controllers/telemetry/tracepipeline_controller.go#L60). This has the advantage that also changes on owned resources without the controller flag trigger a reconciliation.

File: internal/kubernetes/utils.go [hunk 2]
```diff
@@ -189,10 +189,53 @@ func CreateOrUpdateValidatingWebhookConfiguration(ctx context.Context, c client.
 func mergeMetadata(new *metav1.ObjectMeta, old metav1.ObjectMeta) {
 	new.ResourceVersion = old.ResourceVersion
 
+	new.OwnerReferences = restrictControllers(new.OwnerReferences)
```
Comment (chrkl) [inline internal/kubernetes/utils.go:192]: I took over the functionality in a similar way in [this PR](https://github.com/kyma-project/telemetry-manager/pull/127). Revert it here.

File: internal/reconciler/logpipeline/reconciler.go [hunk 3]
```diff
@@ -153,54 +156,88 @@ func (r *Reconciler) doReconcile(ctx context.Context, pipeline *telemetryv1alpha
...
-		return fmt.Errorf("failed to check if LogPipeline is last marked for deletion: %v", err)
-	}
-
-	if shouldDeleteFluentBit {
-		return deleteFluentBit(ctx, r, name)
-	}
-
 	serviceAccount := commonresources.MakeServiceAccount(name)
-	if err = utils.CreateOrUpdateServiceAccount(ctx, r, serviceAccount); err != nil {
+	if err := setOwnerReference(pipeline, serviceAccount, r.Scheme()); err != nil {
```
Comment (chrkl) [inline internal/reconciler/logpipeline/reconciler.go:160]: Let's revert this changes. We can set the owner flag here and clear it on other owners while merging the metadata.

File: internal/kubernetes/utils.go [hunk 4]
```diff
@@ -194,6 +200,81 @@ func mergeMetadata(new *metav1.ObjectMeta, old metav1.ObjectMeta) {
...
+		},
+		GenericFunc: func(evt event.GenericEvent, q workqueue.RateLimitingInterface) {
+			enqueueRequestsForOwners(log, evt.Object, q)
+		},
+		UpdateFunc: func(e event.UpdateEvent, q workqueue.RateLimitingInterface) {
+			oldOwners := getOwnersFromObject(e.ObjectOld)
+			newOwners := getOwnersFromObject(e.ObjectNew)
+
+			for _, newOwner := range newOwners {
+				if ownerSliceContains(oldOwners, newOwner) && hasController(newOwners) {
```
Comment (chrkl) [inline internal/kubernetes/utils.go:219]: I guess we have the possibility that an object has no controller and we would not reconcile. Maybe trigger reconciliation for the first owner in this case.

**Review by chrkl — APPROVED** [review]
